targets/TARGET_STM/mbed_crc_api.c

```diff
@@ -7,50 +7,103 @@
         return false;
     }
 
-    if (config->polynomial != POLY_32BIT_ANSI) {
+#if defined(TARGET_STM32F1) || defined(TARGET_STM32F2) || defined(TARGET_STM32F4) || defined(TARGET_STM32L1)
+    if (config->width != HAL_CRC_LENGTH_32B || config->polynomial != POLY_32BIT_ANSI ||
+        config->initial_xor != 0xFFFFFFFF || config->reflect_in || config->reflect_out) {
         return false;
     }
-
-    if (config->width != 32) {
+#elif defined(TARGET_STM32F0)
+    if (config->width != HAL_CRC_LENGTH_32B || config->polynomial != POLY_32BIT_ANSI) {
         return false;
     }
-
-    if ((config->final_xor != 0xFFFFFFFFU) && (config->final_xor != 0)) {
+#else
+    /* TARGET_STM32L0, TARGET_STM32F3, TARGET_STM32L4, TARGET_STM32F7 */
+    if (config->width != HAL_CRC_LENGTH_32B && config->width != HAL_CRC_LENGTH_16B &&
+        config->width != HAL_CRC_LENGTH_8B && config->width != HAL_CRC_LENGTH_7B) {
         return false;
     }
+#endif
 
     return true;
 }
 
+static uint32_t get_crc_mask(int width)
+{
+    return (width < 8 ? ((1u << 8) - 1) : (uint32_t)((uint64_t)(1ull << width) - 1));
+}
+
 void hal_crc_compute_partial_start(const crc_mbed_config_t *config)
 {
     MBED_ASSERT(hal_crc_is_supported(config));
 
     __HAL_RCC_CRC_CLK_ENABLE();
 
     final_xor = config->final_xor;
+    crc_mask = get_crc_mask(config->width);
 
     current_state.Instance                     = CRC;
+#if !defined(TARGET_STM32F1) && !defined(TARGET_STM32F2) && !defined(TARGET_STM32F4) && !defined(TARGET_STM32L1)
     current_state.InputDataFormat              = CRC_INPUTDATA_FORMAT_BYTES;
-    current_state.Init.DefaultPolynomialUse    = DEFAULT_POLYNOMIAL_ENABLE;
+    current_state.Init.DefaultPolynomialUse    = DEFAULT_POLYNOMIAL_DISABLE;
     current_state.Init.DefaultInitValueUse     = DEFAULT_INIT_VALUE_DISABLE;
     current_state.Init.InitValue               = config->initial_xor;
-    current_state.Init.CRCLength               = CRC_POLYLENGTH_32B;
+    current_state.Init.GeneratingPolynomial    = config->polynomial;
+
+	switch (config->width)
+	{
+	case HAL_CRC_LENGTH_32B:
+		current_state.Init.CRCLength = CRC_POLYLENGTH_32B;
+		break;
+	case HAL_CRC_LENGTH_16B:
+		current_state.Init.CRCLength = CRC_POLYLENGTH_16B;
+		break;
+	case HAL_CRC_LENGTH_8B:
+		current_state.Init.CRCLength = CRC_POLYLENGTH_8B;
+		break;
+	case HAL_CRC_LENGTH_7B:
+		current_state.Init.CRCLength = CRC_POLYLENGTH_7B;
+		break;
+	default:
+		MBED_ASSERT(false);
+		break;
+	}
+
     current_state.Init.InputDataInversionMode  =
         config->reflect_in ? CRC_INPUTDATA_INVERSION_BYTE
         : CRC_INPUTDATA_INVERSION_NONE;
     current_state.Init.OutputDataInversionMode =
         config->reflect_out ? CRC_OUTPUTDATA_INVERSION_ENABLE
         : CRC_OUTPUTDATA_INVERSION_DISABLE;
+#endif
 
-    HAL_CRC_Init(&current_state);
+	if (HAL_CRC_Init(&current_state) != HAL_OK)	{
+		MBED_ASSERT(false);
+	}
 }
 
 void hal_crc_compute_partial(const uint8_t *data, const size_t size)
```

Comment:
I think that is OK, because MbedCRC will use table or bitwise to calculate CRC if hal_crc_is_supported return false.

And I also check `HAL_CRC_Init` in F4, it returns `HAL_OK` always.

Reply:
Sorry, I see the wrong line... let me check first

Reply:
```
/**
  * @brief  Computes the 32-bit CRC of 32-bit data buffer using combination
  *         of the previous CRC value and the new one.
  * @param  hcrc pointer to a CRC_HandleTypeDef structure that contains
  *         the configuration information for CRC
  * @param  pBuffer pointer to the buffer containing the data to be computed
  * @param  BufferLength length of the buffer to be computed
  * @retval 32-bit CRC
  */
uint32_t HAL_CRC_Accumulate(CRC_HandleTypeDef *hcrc, uint32_t pBuffer[], uint32_t BufferLength)
```
The partial of `stm32f4xx_hal_crc.c` as above shows that use 4 bytes alignment only...
and that is very danger when data pointer is not 4 bytes address alignment when optimize option use -O2.

Reply:
I decide set F1/F2/F4/L1 to not supported, more detail please see https://github.com/ARMmbed/mbed-os/pull/9347/commits/ba25cd4e1f83e71f629ca4a67b4580abb5900b67